src/Core/Services/Implementations/CipherService.cs

```diff
@@ -314,6 +323,58 @@ public class CipherService : ICipherService
             }
         }
 
+        public async Task<bool> ValidateCipherAttachmentFile(Cipher cipher, CipherAttachment.MetaData attachmentData)
+        {
+            // var attachmentData = JsonConvert.DeserializeObject<CipherAttachment.MetaData>(attachment.AttachmentData);
+
+            var (valid, realSize) = await _attachmentStorageService.ValidateFileAsync(cipher, attachmentData, _fileSizeLeeway);
+
+            if (!valid || realSize > MAX_FILE_SIZE)
+            {
+                // File reported differs in size from that promised. Must be a rogue client. Delete Send
+                await DeleteAttachmentAsync(cipher, attachmentData.AttachmentId);
+                return false;
+            }
+            // Update Send data if necessary
+            if (realSize != attachmentData.Size)
+            {
+                attachmentData.Size = realSize.Value;
+            }
+            attachmentData.Validated = true;
+
+            var updatedAttachment = new CipherAttachment
+            {
+                Id = cipher.Id,
+                UserId = cipher.UserId,
+                OrganizationId = cipher.OrganizationId,
+                AttachmentId = attachmentData.AttachmentId,
+                AttachmentData = JsonConvert.SerializeObject(attachmentData)
+            };
+
+
+            await _cipherRepository.UpdateAttachmentAsync(updatedAttachment);
+
+            return valid;
+        }
+
+        public async Task<AttachmentResponseModel> GetAttachmentDownloadDataAsync(Cipher cipher, string attachmentId)
+        {
+            var attachments = cipher?.GetAttachments() ?? new Dictionary<string, CipherAttachment.MetaData>();
+
+            if (!attachments.ContainsKey(attachmentId) || attachments[attachmentId].Validated)
```

Comment:
I'm probably misunderstanding what `Validated` property represents, but reads a little backwards in this conditional logic here.

Reply:
The purpose of the `Validated` property is to track whether a blob has been written and validated. If it has, we no longer allow edits to that blob. It's kind of a home-grown write-once system. 

If already written, do not give out a new URL